test/Test/Cardano/Prelude/GHC/Heap/Size.hs

```diff
@@ -0,0 +1,128 @@
+
+-- | Size of a word (in bytes)
+wordSize :: Int
+wordSize = finiteBitSize (0 :: Word) `div` 8
+
+-- | Convert number of bytes to number of machine words
+bytesToNumWords :: Word64 -> NumWords
+bytesToNumWords n = NumWords (fromIntegral n `divRoundUp` wordSize)
+
+-- | Divide @n@ by @m@, rounding up if necessary
+--
+-- > divRoundUp 10 2 == 5
+-- > divRoundUp 10 3 == 4
+divRoundUp :: Int -> Int -> Int
+divRoundUp n m = ceiling (n' / m')
+  where
+    n', m' :: Double
+    n' = fromIntegral n
+    m' = fromIntegral m
+
+-- | Estimate the size of a strict bytestring
+--
+-- Strict 'ByteString' is defined as
+--
+-- > data ByteString = PS {-# UNPACK #-} !(ForeignPtr Word8) -- payload
+-- >                      {-# UNPACK #-} !Int                -- offset
+-- >                      {-# UNPACK #-} !Int                -- length
+--
+-- where
+--
+-- > data ForeignPtr a = ForeignPtr Addr# ForeignPtrContents
+-- > data ForeignPtrContents
+-- >   = PlainForeignPtr !(IORef Finalizers)
+-- >   | MallocPtr      (MutableByteArray# RealWorld) !(IORef Finalizers)
+-- >   | PlainPtr       (MutableByteArray# RealWorld)
+--
+-- The UNPACK pragma on the 'ForeignPtr' means that the @Addr#@ from
+-- 'ForeignPtr' will be inlined in the 'PS' constructor, resulting in a 5-word
+-- size for 'ByteString' and a 2-word size for 'ForeignPtr' (including the info
+-- table pointer). For regularly constructed bytestring only uses 'PlainPtr'.
+--
+-- Since the payload itself will be allocated rounded to the nearest word, a
+-- bytestring of @n@ elements is @5 + 2 + 2 + ceiling (n / wordSize)@ words.
+bsSize :: Int -> NumWords
+bsSize numElems = NumWords (5 + 2 + 2 + numElems `divRoundUp` wordSize)
+
+-- | Check that the reported size is within the specified bounds
+--
+-- Bounds are in words.
+verifySize :: NumWords -> a -> Property
+verifySize expected !x = withTests 1 $ property $ do
+    annotate (show wordSize)
+    sz <- liftIO $ bytesToNumWords <$> computeHeapSize x
+    sz === expected
+
+-- | Wrapper around 'BS.Strict.pack' which cannot be inlined, to avoid
+-- some parts of bytestrings being duplicated rather than shared
+mkBS :: [Word8] -> ByteString
+{-# NOINLINE mkBS #-}
+mkBS = BS.Strict.pack
+
+{-------------------------------------------------------------------------------
+  Tests proper
+
+  We avoid using Double, as that will fit into a single machine word on 64-bit
+  machines but not on 32-bit machines; the bytestring tests check boundary
+  cases around the payload, as well as what happens when we /share/ bytestrings
+  in a tuple.
+
+  For now these are unit tests not property tests; setting up property tests
+  would be a fairly delicate task as we'd have to know a lot about type
+  internals as well as the precise sharing.
+-------------------------------------------------------------------------------}
+
+prop_Int     = verifySize 2 (1    :: Int)
+prop_Float   = verifySize 2 (1.0  :: Float)
+prop_Bool    = verifySize 2 (True :: Bool)
```

Comment:
Is it worth noting why a nullary constructor consumes 2 words? One might expect it to be 1 word, but I believe these are allocated as `CONSTR_0_1`s.

Reply:
Thanks @dcoutts for pointing this out to me. There's a minimum closure size: https://mail.haskell.org/pipermail/ghc-devs/2019-February/017030.html